python/ray/_private/state_api_test_utils.py

```diff
@@ -0,0 +1,284 @@
+            - wait_after_stop: When true, call to `ray.get(actor.stop.remote())`
+                will wait for all pending state APIs to return.
+                Setting it to `False` might miss some long-running state apis calls.
+        """
+        # Configs
+        self._apis = apis
+        self._call_interval_s = call_interval_s
+        self._print_interval_s = print_interval_s
+        self._wait_after_cancel = wait_after_stop
+        self._logger = logging.getLogger(self.__class__.__name__)
+
+        # States
+        self._tasks = None
+        self._fut_queue = None
+        self._executor = None
+        self._loop = None
+        self._stopping = False
+        self._stopped = False
+        self._stats = StateAPIStats()
+
+    async def start(self):
+        # Run the periodic api generator
+        self._fut_queue = asyncio.Queue()
+        self._executor = concurrent.futures.ThreadPoolExecutor()
+
+        self._tasks = [
+            asyncio.ensure_future(awt)
+            for awt in [
+                self._run_generator(),
+                self._run_result_waiter(),
+                self._run_stats_reporter(),
+            ]
+        ]
+        await asyncio.gather(*self._tasks)
+
+    def call(self, fn, verify_cb, **kwargs):
+        def run_fn():
+            try:
+                self._logger.debug(f"calling {fn.__name__}({kwargs})")
+                return invoke_state_api(
+                    verify_cb, fn, state_stats=self._stats, **kwargs
+                )
+            except Exception as e:
+                self._logger.warning(f"{fn.__name__}({kwargs}) failed with: {repr(e)}")
+                return None
+
+        fut = asyncio.get_running_loop().run_in_executor(self._executor, run_fn)
+        return fut
+
+    async def _run_stats_reporter(self):
+        while not self._stopped:
+            # Keep the reporter running until all pending apis finish and the bool
+            # `self._stopped` is then True
+            self._logger.info(pprint.pprint(aggregate_perf_results(self._stats)))
+            try:
+                await asyncio.sleep(self._print_interval_s)
+            except asyncio.CancelledError:
+                self._logger.info(
+                    "_run_stats_reporter cancelled, "
+                    f"waiting for all api {self._stats.pending_calls}calls to return..."
+                )
+
+    async def _run_generator(self):
+        try:
+            while not self._stopping:
+                # Run the state API in another thread
+                for api_spec in self._apis:
+                    fut = self.call(api_spec.api, api_spec.verify_cb, **api_spec.kwargs)
+                    self._fut_queue.put_nowait(fut)
+
+                await asyncio.sleep(self._call_interval_s)
+        except asyncio.CancelledError:
+            # Stop running
+            self._logger.info("_run_generator cancelled, now stopping...")
+            return
+
+    async def _run_result_waiter(self):
```

Comment:
Feel like this class is too complicated. Any reason why it can't just invoke state APIs in the loop (and we can forcefully kill it without handling exit cases?)?

Reply:
> why it can't just invoke state APIs in the loop
Our APIs are currently all blocking, so invoking APIs in a loop will block the loop. A simpler alternative would be using a threaded actor, but since most of the blocking happens at network, I feel it fits the async model better. 

Most of the complexity comes from running the API in another async thread to prevent blocking. If we have async APIs, it will be much easier. 

I also think "invoking a blocking call periodically but async waiting for its results" could be a generic pattern so I went for some over-engineering here.